File: gcj2009QRA.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <regex>

static auto solve = [](const std::vector<std::string>& d)
{
    std::string w; std::cin >> w;
    for(auto& v: w)
        switch(v)
        {
        case '(': v='['; break;
        case ')': v=']'; break;
        default: break;
        }
    std::regex r(w);
    int c=0;
    for(auto& v: d)
        if(std::regex_match(v,r))
            ++c;
    return c;
};

int main()
{
    int L,D,N; std::cin >> L >> D >> N;
    std::vector<std::string> d(D);
    for(auto& v: d) std::cin >> v;
    for(int i=1; i<=N; ++i)
        std::cout << "Case #" << i << ": " << solve(d) << '\n';
}

File: gcj2009QRA.hs
#!/usr/bin/env runghc
import Text.Regex
import Data.Maybe
import Control.Arrow
main = mapM_ putStrLn . zipWith caseno [1..] . solve . lines =<< getContents
caseno i r = "Case #"++show i++": "++r
toString = show
solve (x:xs) = map (toString . answer) ts where
  [l,d,n] = map read $ words x
  (ws,ts) = second (take n) $ splitAt d xs
  answer t = length . flip filter ws $ isJust . matchRegex (cnvRegex t)
cnvRegex = mkRegex . map cnvChar
cnvChar '(' = '['
cnvChar ')' = ']'
cnvChar x = x
